src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+{
+    struct sol_str_slice query, *token;
+    char *i = partial_uri, *sep;
+    uint16_t j;
+    struct sol_vector tokens;
+    int r = 0;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '?')
+        goto exit;
+
+    query.data = ++i;
+
+    for (; *i && *i != '#'; i++) {
+        if (*i != '/' && *i != '?' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the query. "
+                "Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    query.len = i - query.data;
+
+    tokens = sol_util_str_split(query, "&", 0);
+
+    SOL_VECTOR_FOREACH_IDX (&tokens, token, j) {
+        struct sol_http_param_value param;
+        sep = memchr(token->data, '=', token->len);
+        if (!sep) {
+            SOL_ERR("The param: %.*s has no \"=\" separator",
+                SOL_STR_SLICE_PRINT(*token));
+            r = -EINVAL;
+            break;
+        }
+
+        param.value.key_value.key.data = token->data;
+        param.value.key_value.key.len = sep - token->data;
+
+        param.value.key_value.value.data = sep + 1;
+        param.value.key_value.value.len = token->len -
+            param.value.key_value.key.len - 1;
+
+        SOL_DBG("Query key: *%.*s* Query value: *%.*s*",
+            SOL_STR_SLICE_PRINT(param.value.key_value.key),
+            SOL_STR_SLICE_PRINT(param.value.key_value.value));
+
+        param.type = 0;
+
+        if (!sol_http_param_add(query_params, param)) {
+            SOL_ERR("Could not alloc the param %.*s : %.*s",
+                SOL_STR_SLICE_PRINT(param.value.key_value.key),
+                SOL_STR_SLICE_PRINT(param.value.key_value.value));
+            r = -ENOMEM;
+            break;
+        }
+    }
+
+    sol_vector_clear(&tokens);
+exit:
+    *next = i;
+    return r;
+}
+
+static int
+_get_fragment(char *partial_uri, struct sol_str_slice *fragment)
+{
+    struct sol_str_slice f = SOL_STR_SLICE_EMPTY;
+    char *i = partial_uri;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '#') {
+        SOL_ERR("A \"#\" is required in order to identify the fragment");
+        return -EINVAL;
+    }
+
+    f.data = ++i;
+    for (; *i; i++) {
+        if (*i != '/' && *i != '?' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the fragement. "
+                "Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    f.len = i - f.data;
+
+exit:
+    SOL_DBG("URI Fragment: %.*s", SOL_STR_SLICE_PRINT(f));
+    if (fragment)
+        *fragment = f;
+    return 0;
+}
+
+SOL_API int
+sol_http_split_uri(const char *uri, struct sol_str_slice *protocol,
+    struct sol_str_slice *server, struct sol_str_slice *path,
+    struct sol_str_slice *fragment, struct sol_http_param *query_params,
+    int *port)
+{
+    char *next;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+
+    SOL_DBG("Splitting URI: %s", uri);
+
+    r = _get_protocol(uri, protocol, &next);
+    SOL_INT_CHECK(r, < 0, r);
+    r = _get_server(next, server, &next);
```

Comment:
we're not handling local paths ("file:/xx" --- no server), we're not handling userinfo ("http://user@server" or "http://user:pass@server" -- these should be extra parameters to the split function (sorry, I said something else in person, but I had to check the RFC and it's written in there, see https://tools.ietf.org/html/rfc3986#section-3).

As for the implementation, my suggested approach is to have these internal functions to always take and return a slice. 

```c
   struct sol_str_slice partial_uri;

   partial_uri = uri;
   r = _get_scheme(partial_uri, &partial_uri, scheme); // rename protocol to scheme to be close to RFC
   SOL_INT_CHECK(r, < 0, r);
   r = _get_hier_part(partial_uri, &partial_uri, user, password, server, port);
   SOL_INT_CHECK(r, < 0, r);
   r = _get_query(partial_uri, &partial_uri, query);
   if (r != -ENOENT) { // it may not exist, so don't bail... or don't return error, your choice
       SOL_INT_CHECK(r, < 0, r);
   }   
   r = _get_fragment(partial_uri, &partial_uri, fragment);
   if (r != -ENOENT) { // it may not exist, so don't bail... or don't return error, your choice
       SOL_INT_CHECK(r, < 0, r);
   }   
```

Please check with all combinations or URI, add tests to it. Things like:
 * foo://user:pass@server.com:123/a/b?p=1&c=2#/a/b
 * foo://user@server.com:123/a/b?p=1&c=2#/a/b
 * foo://server.com:123/a/b?p=1&c=2#/a/b
 * foo://server.com/a/b?p=1&c=2#/a/b
 * foo://server.com/?p=1&c=2#/a/b (path "/")
 * foo://server.com/?p=1&c=2 (path "/")
 * foo://server.com/#/a/b
 * foo://server.com?p=1&c=2 (path empty)
 * foo://server.com#/a/b (path empty)
 * foo:/a/b?p=1&c=2#/a/b
 * foo:/a/b?p=1&c=2
 * foo:/a/b#/a/b
 * foo:#a/b (path empty)
 * foo:?a/b (path empty)
 * mailto:user@server.com (path "user@server.com")
 * /a/b (fail no scheme)
 * //a/b (fail no scheme)
 * //user:pass@server.com:123/a/b?p=1&c=2#/a/b (fail no scheme)
 * foo://user:pass@server.com:123/a/b?p=1&c=2#/a/b
...

there may be a public list of uris, both positive and negative, to be tested. See http://greenbytes.de/tech/tc/uris/ or http://www.w3.org/wiki/UriTesting.